rama-http/src/service/web/router.rs

```diff
@@ -0,0 +1,440 @@
+use std::convert::Infallible;
+
+use crate::{
+    Request, Response,
+    matcher::{HttpMatcher, UriParams},
+};
+
+use matchit::Router as MatchitRouter;
+use rama_core::{
+    Context,
+    context::Extensions,
+    matcher::Matcher,
+    service::{BoxService, Service},
+};
+use rama_http_types::{Body, StatusCode};
+
+use super::IntoEndpointService;
+
+/// A basic router that can be used to route requests to different services based on the request path.
+///
+/// This router uses `matchit::Router` to efficiently match incoming requests
+/// to predefined routes. Each route is associated with an `HttpMatcher`
+/// and a corresponding service handler.
+pub struct Router<State> {
+    routes: MatchitRouter<
+        Vec<(
+            HttpMatcher<State, Body>,
+            BoxService<State, Request, Response, Infallible>,
+        )>,
+    >,
+    not_found: Option<BoxService<State, Request, Response, Infallible>>,
+}
+
+impl<State> std::fmt::Debug for Router<State> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Router").finish()
+    }
+}
+
+impl<State> Router<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    /// create a new router.
+    pub fn new() -> Self {
+        Self {
+            routes: MatchitRouter::new(),
+            not_found: None,
+        }
+    }
+
+    /// add a GET route to the router.
+    /// the path can contain parameters, e.g. `/users/{id}`.
+    /// the path can also contain a glob matcher, e.g. `/assets/{*path}`.
+    pub fn get<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        // keep everything before the "{*" part and add a glob matcher
+        let matcher_path = if let Some(start) = path.find("{*") {
+            format!("{}*", &path[..start])
+        } else {
+            path.to_owned()
+        };
+
+        let matcher = HttpMatcher::get(matcher_path);
```

Comment:
I get why you did it, but it's a bit silly to also match on path in the matcher given matchit already will ensure we match correctly on path, including nested wise.

So please change these matchers to <https://docs.rs/rama/latest/rama/http/matcher/struct.HttpMatcher.html#method.method> instead.